File: examples/dangling.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char* case1() {
   char x[10]; 
   strcpy(x,"hello");
   printf("\ncase 1 %s", x);
   return x;
}

char *case2() {
  char *x = (char*) malloc(10);
  strcpy(x,"hello");
  printf("\ncase 2 %s", x);
  free(x);
  return x;
}

int main() {
  printf("\nafter case1() %s", case1());
  printf("\nafter case2() %s", case2());
  return 0;
}

File: examples/dangling.rs
#[cfg(feature = "broken")]
fn case1() -> &String {
   let x = String::from("hello"); 
   println!("case 1 {}", x);
   &x 
}

// Below case is not possible with Rust.
// Rust frees the memory automatically when
// the owner goes out of scope and hence
// rust is devoid of such memory freeing errors.
// char *case2() {
// char *x = (char*) malloc(10);
// strcpy(x,"hello");
// free(x);
// printf("\ncase 2 %s", x);
// return x;
// }

fn main() {
  #[cfg(feature = "broken")]
  println!("after case1() {}", case1());
}
